enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,511 @@
+status:
+  conditions:
+  ...
+```
+
+And sets up ACL/RBAC to control who can list/get such objects.
+
+```yaml
+rules:
+- apiGroups:
+  - projected-resource.storage.openshift.io
+  resources:
+  - projectedResources
+  resourceNames:
+  - shared-cool-secret
+  verbs:
+  - get
+  - list
+  - watch
+
+```
+
+A user queries this new type to see what is available to her/him.
+
+Then, to opt in, a user created Pod can cite a CSI volume references (see below for a more detailed explanation of them) that 
+references our new type:
+
+```yaml
+csi:
+  driver: projectedresources.storage.openshift.io
+    volumeAttributes:
+      projectedResourceName: the-projected-resource
+
+```
+
+For pre-existing objects that encapsulate a Pod, a new, documented annotation can be declared to facilitate injection
+of the CSI driver volume specification 
+
+A new admission plugin introduced as part of this feature can then:
+- mutate Pods with this annotation, adding the CSI driver volume specification, if the Pod `ServiceAccount` is allowed sufficient 
+access the `ProjectedResource` (via a SAR check)
+- if a Pod already happens to have the CSI Volume specified with reference to a `ProjectedResource`, again see if the 
+associated `ServiceAccount` for the Pod has sufficient access to the `ProjectedResource`
+- if sufficient access does not exist, the admission plugin marks the Pod as invalid/forbidden.
+- if sufficient access does exists, the `volumeAttributes` are also updated to add a key/value pair for the requesting
+`ServiceAccont` (including both its name and namespace)
+
+```yaml
+csi:
+  driver: projectedresources.storage.openshift.io
+    volumeAttributes:
+      projectedResourceName: the-projected-resource
+      requestingNamespace: the-users-namespace
+      requestingServiceAccount: the-sa-the-user-is-using-for-this-object
+```
+
+The CSI Driver, aside from all the CSI and filesystem bits (see below for details), performs:
+- reads of the `ProjectedResource` object referenced in the `volumeMetadata`
+- reads of the `Secret` or `ConfigMap` noted in the `ProjectedResource`, storing it on disk on initial
+reference to said `Secret` or `ConfigMap`
+- creation of a tmpfs filesystem for each requesting `ServiceAcount` that contains the `Secret` / `ConfigMap` data
+- creation of k8s watches `ProjectedResources`, and for `Secret` / `ConfigMaps` in the namespaces listed in each discovered
+`ProjectedResources`
+- final returning of a volume that is based off of the `ServiceAccount` specific link for the data noted in the 
+`volumeAttributes`
+
+After initial set up of things, steady state CSI driver activities include:
+- as updates of the watched `Secrets` and `ConfigMaps` come in, SAR checks will be done for each of the previous requesting 
+`ServiceAccounts` being tracked against the updated `Secret` and `ConfigMap`
+- for any `ServiceAccounts` that no longer have access, their tmpfs filesystems are unmounted, effectively removing their access to 
+the data
```

Comment:
Not sure this is going to work due to mount propagation. It's probably OK to leave it as it is now, but test this carefully.

Reply:
I'll add a comment that we need to keep an eye on this thanks